File: Traditional-Important-Algorithms/exercises_py/06_distributed_systems/dist02_load_balancing.py
# I AM NOT DONE

"""
dist02_load_balancing.py

Load Balancing distributes incoming requests across multiple servers to
optimize resource utilization, minimize response time, and avoid overload.

Common algorithms:
- Round Robin: Distribute requests sequentially across servers
- Weighted Round Robin: Assign more requests to more capable servers
- Least Connections: Route to server with fewest active connections
- Weighted Least Connections: Combine connection count with server capacity

Your task: Implement multiple load balancing strategies.

Key concepts:
- Server health and availability
- Connection tracking
- Fair distribution vs. optimal distribution
- Server weights/capacity
"""

from dataclasses import dataclass
from enum import Enum, auto
from typing import Dict, List, Optional
import unittest


class LoadBalancingStrategy(Enum):
    """Load balancing strategy types."""
    ROUND_ROBIN = auto()
    WEIGHTED_ROUND_ROBIN = auto()
    LEAST_CONNECTIONS = auto()
    WEIGHTED_LEAST_CONNECTIONS = auto()


@dataclass
class Server:
    """Server representation."""
    id: int
    weight: int
    active_connections: int = 0
    total_requests: int = 0
    is_healthy: bool = True

    def add_connection(self):
        """Add a connection to this server."""
        self.active_connections += 1
        self.total_requests += 1

    def remove_connection(self):
        """Remove a connection from this server."""
        if self.active_connections > 0:
            self.active_connections -= 1


@dataclass
class ServerStats:
    """Statistics for a server."""
    server_id: int
    weight: int
    active_connections: int
    total_requests: int
    is_healthy: bool


class LoadBalancer:
    """Load balancer implementing multiple strategies."""

    def __init__(self, strategy: LoadBalancingStrategy):
        """Initialize load balancer with a strategy."""
        self.servers: List[Server] = []
        self.strategy = strategy
        self.round_robin_index = 0
        self.weighted_rr_counter: Dict[int, int] = {}

    def add_server(self, server: Server):
        """Add a server to the load balancer."""
        self.weighted_rr_counter[server.id] = 0
        self.servers.append(server)

    def mark_server_unhealthy(self, server_id: int):
        """Mark a server as unhealthy."""
        for server in self.servers:
            if server.id == server_id:
                server.is_healthy = False
                break

    def mark_server_healthy(self, server_id: int):
        """Mark a server as healthy."""
        for server in self.servers:
            if server.id == server_id:
                server.is_healthy = True
                break

    def _get_healthy_servers(self) -> List[int]:
        """Get indices of healthy servers."""
        return [i for i, s in enumerate(self.servers) if s.is_healthy]

    def _select_round_robin(self) -> Optional[int]:
        """
        TODO: Implement round robin selection.

        - Find next healthy server starting from round_robin_index
        - Update round_robin_index for next call
        - Return server ID (not index)
        - Return None if no healthy servers
        """
        pass  # TODO: Implement this

    def _select_weighted_round_robin(self) -> Optional[int]:
        """
        TODO: Implement weighted round robin selection.

        - Each server should receive requests proportional to its weight
        - For example, servers with weights [1, 2, 3] should receive
          requests in ratio 1:2:3
        - Use weighted_rr_counter to track distribution
        - Return server ID (not index)
        - Return None if no healthy servers
        """
        pass  # TODO: Implement this

    def _select_least_connections(self) -> Optional[int]:
        """
        TODO: Implement least connections selection.

        - Find healthy server with minimum active_connections
        - If tie, choose the one with lower ID
        - Return server ID (not index)
        - Return None if no healthy servers
        """
        pass  # TODO: Implement this

    def _select_weighted_least_connections(self) -> Optional[int]:
        """
        TODO: Implement weighted least connections selection.

        - Find healthy server with minimum (active_connections / weight) ratio
        - This balances load relative to server capacity
        - If tie, choose the one with lower ID
        - Return server ID (not index)
        - Return None if no healthy servers
        """
        pass  # TODO: Implement this

    def select_server(self) -> Optional[int]:
        """Select a server based on the configured strategy."""
        if self.strategy == LoadBalancingStrategy.ROUND_ROBIN:
            return self._select_round_robin()
        elif self.strategy == LoadBalancingStrategy.WEIGHTED_ROUND_ROBIN:
            return self._select_weighted_round_robin()
        elif self.strategy == LoadBalancingStrategy.LEAST_CONNECTIONS:
            return self._select_least_connections()
        elif self.strategy == LoadBalancingStrategy.WEIGHTED_LEAST_CONNECTIONS:
            return self._select_weighted_least_connections()

    def handle_request(self) -> Optional[int]:
        """
        TODO: Select a server and add a connection to it.

        Return the server ID.
        """
        pass  # TODO: Implement this

    def complete_request(self, server_id: int):
        """
        TODO: Remove a connection from the specified server.
        """
        pass  # TODO: Implement this

    def get_server(self, server_id: int) -> Optional[Server]:
        """Get a server by ID."""
        for server in self.servers:
            if server.id == server_id:
                return server
        return None

    def get_server_stats(self) -> List[ServerStats]:
        """Get statistics for all servers."""
        return [
            ServerStats(
                server_id=s.id,
                weight=s.weight,
                active_connections=s.active_connections,
                total_requests=s.total_requests,
                is_healthy=s.is_healthy
            )
            for s in self.servers
        ]


# Unit Tests
class TestLoadBalancing(unittest.TestCase):

    def test_round_robin_basic(self):
        lb = LoadBalancer(LoadBalancingStrategy.ROUND_ROBIN)
        lb.add_server(Server(0, 1))
        lb.add_server(Server(1, 1))
        lb.add_server(Server(2, 1))

        # Should cycle through servers
        self.assertEqual(lb.select_server(), 0)
        self.assertEqual(lb.select_server(), 1)
        self.assertEqual(lb.select_server(), 2)
        self.assertEqual(lb.select_server(), 0)

    def test_round_robin_skip_unhealthy(self):
        lb = LoadBalancer(LoadBalancingStrategy.ROUND_ROBIN)
        lb.add_server(Server(0, 1))
        lb.add_server(Server(1, 1))
        lb.add_server(Server(2, 1))

        lb.mark_server_unhealthy(1)

        # Should skip server 1
        self.assertEqual(lb.select_server(), 0)
        self.assertEqual(lb.select_server(), 2)
        self.assertEqual(lb.select_server(), 0)

    def test_weighted_round_robin(self):
        lb = LoadBalancer(LoadBalancingStrategy.WEIGHTED_ROUND_ROBIN)
        lb.add_server(Server(0, 1))
        lb.add_server(Server(1, 2))

        # Collect 12 selections
        selections = []
        for _ in range(12):
            selections.append(lb.select_server())

        # Count occurrences
        count_0 = selections.count(0)
        count_1 = selections.count(1)

        # Should be roughly 1:2 ratio (4:8 in 12 requests)
        self.assertEqual(count_0, 4)
        self.assertEqual(count_1, 8)

    def test_least_connections_basic(self):
        lb = LoadBalancer(LoadBalancingStrategy.LEAST_CONNECTIONS)
        lb.add_server(Server(0, 1))
        lb.add_server(Server(1, 1))

        # First request goes to server 0 (both have 0 connections, choose lower ID)
        server = lb.handle_request()
        self.assertEqual(server, 0)

        # Second request should go to server 1 (has fewer connections)
        server = lb.handle_request()
        self.assertEqual(server, 1)

    def test_least_connections_balancing(self):
        lb = LoadBalancer(LoadBalancingStrategy.LEAST_CONNECTIONS)
        lb.add_server(Server(0, 1))
        lb.add_server(Server(1, 1))

        # Add 2 connections to server 0
        lb.handle_request()
        lb.handle_request()

        # Complete one request from server 0
        lb.complete_request(0)

        # Server 0 has 1 connection, server 1 has 0
        # Next should go to server 1
        server = lb.select_server()
        self.assertEqual(server, 1)

    def test_weighted_least_connections(self):
        lb = LoadBalancer(LoadBalancingStrategy.WEIGHTED_LEAST_CONNECTIONS)
        lb.add_server(Server(0, 1))  # Lower capacity
        lb.add_server(Server(1, 3))  # Higher capacity

        # Send requests
        for _ in range(8):
            lb.handle_request()

        stats = lb.get_server_stats()
        s0 = next(s for s in stats if s.server_id == 0)
        s1 = next(s for s in stats if s.server_id == 1)

        # Server 1 should handle more requests due to higher weight
        self.assertGreater(s1.active_connections, s0.active_connections)

    def test_no_healthy_servers(self):
        lb = LoadBalancer(LoadBalancingStrategy.ROUND_ROBIN)
        lb.add_server(Server(0, 1))
        lb.add_server(Server(1, 1))

        lb.mark_server_unhealthy(0)
        lb.mark_server_unhealthy(1)

        self.assertIsNone(lb.select_server())

    def test_server_recovery(self):
        lb = LoadBalancer(LoadBalancingStrategy.ROUND_ROBIN)
        lb.add_server(Server(0, 1))
        lb.add_server(Server(1, 1))

        lb.mark_server_unhealthy(1)
        self.assertEqual(lb.select_server(), 0)

        lb.mark_server_healthy(1)
        self.assertEqual(lb.select_server(), 0)
        self.assertEqual(lb.select_server(), 1)

    def test_connection_tracking(self):
        lb = LoadBalancer(LoadBalancingStrategy.ROUND_ROBIN)
        lb.add_server(Server(0, 1))

        lb.handle_request()
        lb.handle_request()

        server = lb.get_server(0)
        self.assertEqual(server.active_connections, 2)
        self.assertEqual(server.total_requests, 2)

        lb.complete_request(0)
        server = lb.get_server(0)
        self.assertEqual(server.active_connections, 1)
        self.assertEqual(server.total_requests, 2)  # Total doesn't decrease

    def test_weighted_distribution_fairness(self):
        lb = LoadBalancer(LoadBalancingStrategy.WEIGHTED_ROUND_ROBIN)
        lb.add_server(Server(0, 1))
        lb.add_server(Server(1, 2))
        lb.add_server(Server(2, 3))

        # Send 60 requests (divisible by 1+2+3=6)
        distribution = {}
        for _ in range(60):
            server_id = lb.select_server()
            distribution[server_id] = distribution.get(server_id, 0) + 1

        # Should distribute in ratio 1:2:3 = 10:20:30
        self.assertEqual(distribution.get(0), 10)
        self.assertEqual(distribution.get(1), 20)
        self.assertEqual(distribution.get(2), 30)

    def test_statistics(self):
        lb = LoadBalancer(LoadBalancingStrategy.ROUND_ROBIN)
        lb.add_server(Server(0, 2))
        lb.add_server(Server(1, 3))

        lb.handle_request()
        lb.handle_request()

        stats = lb.get_server_stats()
        self.assertEqual(len(stats), 2)

        s0_stats = next(s for s in stats if s.server_id == 0)
        self.assertEqual(s0_stats.weight, 2)
        self.assertTrue(s0_stats.is_healthy)

    def test_least_connections_tie_breaking(self):
        lb = LoadBalancer(LoadBalancingStrategy.LEAST_CONNECTIONS)
        lb.add_server(Server(5, 1))
        lb.add_server(Server(3, 1))
        lb.add_server(Server(7, 1))

        # All have 0 connections, should choose lowest ID (3)
        server = lb.select_server()
        self.assertEqual(server, 3)


if __name__ == '__main__':
    unittest.main()

File: Traditional-Important-Algorithms/exercises/06_distributed_systems/dist02_load_balancing.rs
// dist02_load_balancing.rs
//
// Load Balancing distributes incoming requests across multiple servers to
// optimize resource utilization, minimize response time, and avoid overload.
//
// Common algorithms:
// - Round Robin: Distribute requests sequentially across servers
// - Weighted Round Robin: Assign more requests to more capable servers
// - Least Connections: Route to server with fewest active connections
// - Weighted Least Connections: Combine connection count with server capacity
//
// Your task: Implement multiple load balancing strategies.
//
// Key concepts:
// - Server health and availability
// - Connection tracking
// - Fair distribution vs. optimal distribution
// - Server weights/capacity

// I AM NOT DONE

use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq)]
pub struct Server {
    pub id: usize,
    pub weight: usize,          // Capacity weight (higher = more capable)
    pub active_connections: usize,
    pub total_requests: usize,
    pub is_healthy: bool,
}

impl Server {
    pub fn new(id: usize, weight: usize) -> Self {
        Self {
            id,
            weight,
            active_connections: 0,
            total_requests: 0,
            is_healthy: true,
        }
    }

    pub fn add_connection(&mut self) {
        self.active_connections += 1;
        self.total_requests += 1;
    }

    pub fn remove_connection(&mut self) {
        if self.active_connections > 0 {
            self.active_connections -= 1;
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum LoadBalancingStrategy {
    RoundRobin,
    WeightedRoundRobin,
    LeastConnections,
    WeightedLeastConnections,
}

pub struct LoadBalancer {
    servers: Vec<Server>,
    strategy: LoadBalancingStrategy,
    round_robin_index: usize,
    weighted_rr_counter: HashMap<usize, usize>, // Track weighted distribution
}

impl LoadBalancer {
    pub fn new(strategy: LoadBalancingStrategy) -> Self {
        Self {
            servers: Vec::new(),
            strategy,
            round_robin_index: 0,
            weighted_rr_counter: HashMap::new(),
        }
    }

    pub fn add_server(&mut self, server: Server) {
        self.weighted_rr_counter.insert(server.id, 0);
        self.servers.push(server);
    }

    pub fn mark_server_unhealthy(&mut self, server_id: usize) {
        if let Some(server) = self.servers.iter_mut().find(|s| s.id == server_id) {
            server.is_healthy = false;
        }
    }

    pub fn mark_server_healthy(&mut self, server_id: usize) {
        if let Some(server) = self.servers.iter_mut().find(|s| s.id == server_id) {
            server.is_healthy = true;
        }
    }

    fn get_healthy_servers(&self) -> Vec<usize> {
        self.servers
            .iter()
            .enumerate()
            .filter(|(_, s)| s.is_healthy)
            .map(|(i, _)| i)
            .collect()
    }

    fn select_round_robin(&mut self) -> Option<usize> {
        // TODO: Implement round robin selection
        // - Find next healthy server starting from round_robin_index
        // - Update round_robin_index for next call
        // - Return server ID (not index)
        // - Return None if no healthy servers
        todo!()
    }

    fn select_weighted_round_robin(&mut self) -> Option<usize> {
        // TODO: Implement weighted round robin selection
        // - Each server should receive requests proportional to its weight
        // - For example, servers with weights [1, 2, 3] should receive
        //   requests in ratio 1:2:3
        // - Use weighted_rr_counter to track distribution
        // - Return server ID (not index)
        // - Return None if no healthy servers
        todo!()
    }

    fn select_least_connections(&self) -> Option<usize> {
        // TODO: Implement least connections selection
        // - Find healthy server with minimum active_connections
        // - If tie, choose the one with lower ID
        // - Return server ID (not index)
        // - Return None if no healthy servers
        todo!()
    }

    fn select_weighted_least_connections(&self) -> Option<usize> {
        // TODO: Implement weighted least connections selection
        // - Find healthy server with minimum (active_connections / weight) ratio
        // - This balances load relative to server capacity
        // - If tie, choose the one with lower ID
        // - Return server ID (not index)
        // - Return None if no healthy servers
        todo!()
    }

    pub fn select_server(&mut self) -> Option<usize> {
        match self.strategy {
            LoadBalancingStrategy::RoundRobin => self.select_round_robin(),
            LoadBalancingStrategy::WeightedRoundRobin => self.select_weighted_round_robin(),
            LoadBalancingStrategy::LeastConnections => self.select_least_connections(),
            LoadBalancingStrategy::WeightedLeastConnections => self.select_weighted_least_connections(),
        }
    }

    pub fn handle_request(&mut self) -> Option<usize> {
        // TODO: Select a server and add a connection to it
        // Return the server ID
        todo!()
    }

    pub fn complete_request(&mut self, server_id: usize) {
        // TODO: Remove a connection from the specified server
        todo!()
    }

    pub fn get_server(&self, server_id: usize) -> Option<&Server> {
        self.servers.iter().find(|s| s.id == server_id)
    }

    pub fn get_server_stats(&self) -> Vec<ServerStats> {
        self.servers
            .iter()
            .map(|s| ServerStats {
                server_id: s.id,
                weight: s.weight,
                active_connections: s.active_connections,
                total_requests: s.total_requests,
                is_healthy: s.is_healthy,
            })
            .collect()
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct ServerStats {
    pub server_id: usize,
    pub weight: usize,
    pub active_connections: usize,
    pub total_requests: usize,
    pub is_healthy: bool,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_round_robin_basic() {
        let mut lb = LoadBalancer::new(LoadBalancingStrategy::RoundRobin);
        lb.add_server(Server::new(0, 1));
        lb.add_server(Server::new(1, 1));
        lb.add_server(Server::new(2, 1));

        // Should cycle through servers
        assert_eq!(lb.select_server(), Some(0));
        assert_eq!(lb.select_server(), Some(1));
        assert_eq!(lb.select_server(), Some(2));
        assert_eq!(lb.select_server(), Some(0));
    }

    #[test]
    fn test_round_robin_skip_unhealthy() {
        let mut lb = LoadBalancer::new(LoadBalancingStrategy::RoundRobin);
        lb.add_server(Server::new(0, 1));
        lb.add_server(Server::new(1, 1));
        lb.add_server(Server::new(2, 1));

        lb.mark_server_unhealthy(1);

        // Should skip server 1
        assert_eq!(lb.select_server(), Some(0));
        assert_eq!(lb.select_server(), Some(2));
        assert_eq!(lb.select_server(), Some(0));
    }

    #[test]
    fn test_weighted_round_robin() {
        let mut lb = LoadBalancer::new(LoadBalancingStrategy::WeightedRoundRobin);
        lb.add_server(Server::new(0, 1));
        lb.add_server(Server::new(1, 2));

        // Collect 12 selections
        let mut selections = Vec::new();
        for _ in 0..12 {
            selections.push(lb.select_server().unwrap());
        }

        // Count occurrences
        let count_0 = selections.iter().filter(|&&id| id == 0).count();
        let count_1 = selections.iter().filter(|&&id| id == 1).count();

        // Should be roughly 1:2 ratio (4:8 in 12 requests)
        assert_eq!(count_0, 4);
        assert_eq!(count_1, 8);
    }

    #[test]
    fn test_least_connections_basic() {
        let mut lb = LoadBalancer::new(LoadBalancingStrategy::LeastConnections);
        lb.add_server(Server::new(0, 1));
        lb.add_server(Server::new(1, 1));

        // First request goes to server 0 (both have 0 connections, choose lower ID)
        let server = lb.handle_request().unwrap();
        assert_eq!(server, 0);

        // Second request should go to server 1 (has fewer connections)
        let server = lb.handle_request().unwrap();
        assert_eq!(server, 1);
    }

    #[test]
    fn test_least_connections_balancing() {
        let mut lb = LoadBalancer::new(LoadBalancingStrategy::LeastConnections);
        lb.add_server(Server::new(0, 1));
        lb.add_server(Server::new(1, 1));

        // Add 2 connections to server 0
        lb.handle_request();
        lb.handle_request();

        // Complete one request from server 0
        lb.complete_request(0);

        // Server 0 has 1 connection, server 1 has 0
        // Next should go to server 1
        let server = lb.select_server().unwrap();
        assert_eq!(server, 1);
    }

    #[test]
    fn test_weighted_least_connections() {
        let mut lb = LoadBalancer::new(LoadBalancingStrategy::WeightedLeastConnections);
        lb.add_server(Server::new(0, 1)); // Lower capacity
        lb.add_server(Server::new(1, 3)); // Higher capacity

        // Send requests
        for _ in 0..8 {
            lb.handle_request();
        }

        let stats = lb.get_server_stats();
        let s0 = stats.iter().find(|s| s.server_id == 0).unwrap();
        let s1 = stats.iter().find(|s| s.server_id == 1).unwrap();

        // Server 1 should handle more requests due to higher weight
        assert!(s1.active_connections > s0.active_connections);
    }

    #[test]
    fn test_no_healthy_servers() {
        let mut lb = LoadBalancer::new(LoadBalancingStrategy::RoundRobin);
        lb.add_server(Server::new(0, 1));
        lb.add_server(Server::new(1, 1));

        lb.mark_server_unhealthy(0);
        lb.mark_server_unhealthy(1);

        assert_eq!(lb.select_server(), None);
    }

    #[test]
    fn test_server_recovery() {
        let mut lb = LoadBalancer::new(LoadBalancingStrategy::RoundRobin);
        lb.add_server(Server::new(0, 1));
        lb.add_server(Server::new(1, 1));

        lb.mark_server_unhealthy(1);
        assert_eq!(lb.select_server(), Some(0));

        lb.mark_server_healthy(1);
        assert_eq!(lb.select_server(), Some(0));
        assert_eq!(lb.select_server(), Some(1));
    }

    #[test]
    fn test_connection_tracking() {
        let mut lb = LoadBalancer::new(LoadBalancingStrategy::RoundRobin);
        lb.add_server(Server::new(0, 1));

        lb.handle_request();
        lb.handle_request();

        let server = lb.get_server(0).unwrap();
        assert_eq!(server.active_connections, 2);
        assert_eq!(server.total_requests, 2);

        lb.complete_request(0);
        let server = lb.get_server(0).unwrap();
        assert_eq!(server.active_connections, 1);
        assert_eq!(server.total_requests, 2); // Total doesn't decrease
    }

    #[test]
    fn test_weighted_distribution_fairness() {
        let mut lb = LoadBalancer::new(LoadBalancingStrategy::WeightedRoundRobin);
        lb.add_server(Server::new(0, 1));
        lb.add_server(Server::new(1, 2));
        lb.add_server(Server::new(2, 3));

        // Send 60 requests (divisible by 1+2+3=6)
        let mut distribution = HashMap::new();
        for _ in 0..60 {
            let server_id = lb.select_server().unwrap();
            *distribution.entry(server_id).or_insert(0) += 1;
        }

        // Should distribute in ratio 1:2:3 = 10:20:30
        assert_eq!(distribution.get(&0), Some(&10));
        assert_eq!(distribution.get(&1), Some(&20));
        assert_eq!(distribution.get(&2), Some(&30));
    }

    #[test]
    fn test_statistics() {
        let mut lb = LoadBalancer::new(LoadBalancingStrategy::RoundRobin);
        lb.add_server(Server::new(0, 2));
        lb.add_server(Server::new(1, 3));

        lb.handle_request();
        lb.handle_request();

        let stats = lb.get_server_stats();
        assert_eq!(stats.len(), 2);

        let s0_stats = stats.iter().find(|s| s.server_id == 0).unwrap();
        assert_eq!(s0_stats.weight, 2);
        assert!(s0_stats.is_healthy);
    }

    #[test]
    fn test_least_connections_tie_breaking() {
        let mut lb = LoadBalancer::new(LoadBalancingStrategy::LeastConnections);
        lb.add_server(Server::new(5, 1));
        lb.add_server(Server::new(3, 1));
        lb.add_server(Server::new(7, 1));

        // All have 0 connections, should choose lowest ID (3)
        let server = lb.select_server().unwrap();
        assert_eq!(server, 3);
    }
}
